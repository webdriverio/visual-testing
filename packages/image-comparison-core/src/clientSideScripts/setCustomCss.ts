import type { CssOptions } from './customCss.interfaces.js'

/**
 * Set some default css
 */
export default function setCustomCss(cssOptions: CssOptions): void {
    if (!document.head) {
        return
    }
    // disabling CSS animations for everything including pseudo elements
    const disableTransformationsTransitionsAnimations = `
*, *::before, *::after {
    -o-transition-property: none !important;
    -moz-transition-property: none !important;
    -ms-transition-property: none !important;
    -webkit-transition-property: none !important;
    transition-property: none !important;
    -webkit-animation: none !important;
    -moz-animation: none !important;
    -o-animation: none !important;
    -ms-animation: none !important;
    animation: none !important;
    caret-color: transparent !important;
}`
    const { addressBarPadding, disableBlinkingCursor, disableCSSAnimation, id, toolBarPadding } = cssOptions
    const bodyTopPadding = addressBarPadding === 0 ? '' : `body{padding-top:${addressBarPadding}px !important}`
    const bodyBottomPadding = toolBarPadding === 0 ? '' : `body{padding-bottom:${toolBarPadding}px !important}`
    const disableBlinkingCursorCss = disableBlinkingCursor ? ' input, textarea, [contenteditable]{caret-color: transparent !important;}' : ''
    const css = (disableCSSAnimation ? disableTransformationsTransitionsAnimations : '') +
        bodyTopPadding +
        bodyBottomPadding +
        disableBlinkingCursorCss
    const head = document.head
    const style = document.createElement('style')

    style.id = id
    style.appendChild(document.createTextNode(css))
    head.appendChild(style)
}
