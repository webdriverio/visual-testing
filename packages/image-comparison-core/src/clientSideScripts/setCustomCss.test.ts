// @vitest-environment jsdom

import { beforeEach, describe, it, expect } from 'vitest'
import type { CssOptions } from './customCss.interfaces.js'
import setCustomCss from './setCustomCss.js'

describe('setCustomCss', () => {
    beforeEach(() => {
        document.head.innerHTML = ''
    })

    it('should be able to set the custom css with the default options', () => {
        const cssOptions: CssOptions = {
            addressBarPadding: 6,
            disableBlinkingCursor: false,
            disableCSSAnimation: false,
            id: 'id',
            toolBarPadding: 6,
        }

        expect(document.head.textContent).toMatchSnapshot()

        setCustomCss(cssOptions)

        expect(document.head.textContent).toMatchSnapshot()
    })

    it('should be able to set the custom css with the animations disabled', () => {
        const cssOptions: CssOptions = {
            addressBarPadding: 6,
            disableBlinkingCursor: false,
            disableCSSAnimation: true,
            id: 'id',
            toolBarPadding: 6,
        }

        expect(document.head.textContent).toMatchSnapshot()

        setCustomCss(cssOptions)

        expect(document.head.textContent).toMatchSnapshot()
    })

    it('should be able to set the custom css with the with padding set to 0', () => {
        const cssOptions: CssOptions = {
            addressBarPadding: 0,
            disableBlinkingCursor: false,
            disableCSSAnimation: true,
            id: 'id',
            toolBarPadding: 0,
        }

        expect(document.head.textContent).toMatchSnapshot()

        setCustomCss(cssOptions)

        expect(document.head.textContent).toMatchSnapshot()
    })

    it('should be able to set the custom css with the with disableBlinkingCursor set to true', () => {
        const cssOptions: CssOptions = {
            addressBarPadding: 0,
            disableBlinkingCursor: true,
            disableCSSAnimation: false,
            id: 'id',
            toolBarPadding: 0,
        }

        expect(document.head.textContent).toMatchSnapshot()

        setCustomCss(cssOptions)

        expect(document.head.textContent).toMatchSnapshot()
    })

    it('should do nothing if document.head is null', () => {
        const cssOptions: CssOptions = {
            addressBarPadding: 6,
            disableBlinkingCursor: false,
            disableCSSAnimation: false,
            id: 'id',
            toolBarPadding: 6,
        }
        Object.defineProperty(document, 'head', { value: null })

        setCustomCss(cssOptions)

        expect(document.head).toBe(null)
    })
})
