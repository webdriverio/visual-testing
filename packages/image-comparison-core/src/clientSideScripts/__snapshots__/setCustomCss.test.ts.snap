// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setCustomCss > should be able to set the custom css with the animations disabled 1`] = `""`;

exports[`setCustomCss > should be able to set the custom css with the animations disabled 2`] = `
"
*, *::before, *::after {
    -o-transition-property: none !important;
    -moz-transition-property: none !important;
    -ms-transition-property: none !important;
    -webkit-transition-property: none !important;
    transition-property: none !important;
    -webkit-animation: none !important;
    -moz-animation: none !important;
    -o-animation: none !important;
    -ms-animation: none !important;
    animation: none !important;
    caret-color: transparent !important;
}body{padding-top:6px !important}body{padding-bottom:6px !important}"
`;

exports[`setCustomCss > should be able to set the custom css with the default options 1`] = `""`;

exports[`setCustomCss > should be able to set the custom css with the default options 2`] = `"body{padding-top:6px !important}body{padding-bottom:6px !important}"`;

exports[`setCustomCss > should be able to set the custom css with the with disableBlinkingCursor set to true 1`] = `""`;

exports[`setCustomCss > should be able to set the custom css with the with disableBlinkingCursor set to true 2`] = `" input, textarea, [contenteditable]{caret-color: transparent !important;}"`;

exports[`setCustomCss > should be able to set the custom css with the with padding set to 0 1`] = `""`;

exports[`setCustomCss > should be able to set the custom css with the with padding set to 0 2`] = `
"
*, *::before, *::after {
    -o-transition-property: none !important;
    -moz-transition-property: none !important;
    -ms-transition-property: none !important;
    -webkit-transition-property: none !important;
    transition-property: none !important;
    -webkit-animation: none !important;
    -moz-animation: none !important;
    -o-animation: none !important;
    -ms-animation: none !important;
    animation: none !important;
    caret-color: transparent !important;
}"
`;
